src/mem/cache/prefetch/ipcp.cc

```diff
@@ -0,0 +1,398 @@
+        ipcpTable[ipIndex].lastLineOffset = lineOffset;
+        ipcpTable[ipIndex].signature = 0;
+        ipcpTable[ipIndex].stride = 0;
+        ipcpTable[ipIndex].conf = 0;
+        //ipcpTable[ipIndex].conf = 4;
+        ipcpTable[ipIndex].valid = 1;
+        ipcpTable[ipIndex].regionAddr = regionBase;
+        ipcpTable[ipIndex].dir = 1;
+        ipcpTable[ipIndex].streamValid = (RST[regionBase].tentative
+         || RST[regionBase].trained);
+    }
+
+    curStride = lineOffset - ipcpTable[ipIndex].lastLineOffset;
+    // Page boundary learning
+    if (ipcpTable[ipIndex].lastVpage != curPage) { //switching it off?
+       if (curStride < 0) //Assuming the page moving in +ve direction
+           curStride += blkSize;
+       else
+           curStride -= blkSize;
+    }
+    if (curStride == 0) return;
+
+    if (ipcpTable[ipIndex].stride == curStride) {
+        ipcpTable[ipIndex].conf += 1;
+    }
+    else
+        // should we return from here?
+        ipcpTable[ipIndex].conf -= 1;
+
+    //Overflow fix
+    if (ipcpTable[ipIndex].conf > 4) ipcpTable[ipIndex].conf = 4;
+    if (ipcpTable[ipIndex].conf < 0) ipcpTable[ipIndex].conf = 0;
+
+    //Update stride
+    if (ipcpTable[ipIndex].conf == 0)
+        ipcpTable[ipIndex].stride = curStride;
+
+    //Update lastPage and lastlineOffset
+    ipcpTable[ipIndex].lastVpage = curPage;
+    ipcpTable[ipIndex].lastLineOffset = lineOffset;
+
+    Addr lastSignature =  ipcpTable[ipIndex].signature;
+
+    //Hit in CSPT
+    if (cspt.find(lastSignature) != cspt.end()) {
+        if (cspt[lastSignature].stride == curStride)
+            cspt[lastSignature].conf += 1;
+        else
+            cspt[lastSignature].conf -= 1;
+        //Overflow fix
+        if (cspt[lastSignature].conf > 4) cspt[lastSignature].conf = 4;
+        if (cspt[lastSignature].conf < 0) cspt[lastSignature].conf = 0;
+
+        //Update stride
+        if (cspt[lastSignature].conf == 0)
+           cspt[lastSignature].stride = curStride;
+
+    } else {
+       cspt[lastSignature].stride = curStride; //this may not be correct
+       cspt[lastSignature].conf = 0;
+    }
+
+    //update next signature
+    Addr updatedSignature = getSignature(lastSignature, curStride);
+    ipcpTable[ipIndex].signature = updatedSignature;
+
+    //Prediction
+    if (pageOn && ipcpTable[ipIndex].streamValid) {
+        ipcpStats.pageChosen += 1;
+        int64_t newStride = ipcpTable[ipIndex].dir * blkSize;
+        if (regionBase != ipcpTable[ipIndex].regionAddr) {
+            if (0 && !RST[regionBase].trained) {
+               RST[regionBase].tentative = true;
+            }
+            ipcpTable[ipIndex].regionAddr = regionBase;
+        }
+        for (int i = 1; i <= pageDegree; i++) {
+            Addr prefAddr = blockAddress(blkAddr + (i * newStride));
+            addresses.push_back(AddrPriority(prefAddr,0));
+            if (!samePage(blkAddr, prefAddr))
+                ipcpStats.pageCross += 1;
```

Comment:
Use curly braces everywhere. It's particularly confusing (and can lead to hard to find bugs!) when sometimes you use curly braces for single line if statements and other times you don't.